### PR TITLE
A cool solution with user-controlled graph traversal pruning for assignment 3

### PR DESCRIPTION
**Pull request by jdb**

### CODE REVIEW

**Comment by evinpinar** [comment]
It's a great solution, thanks!

**Comment by SilverSolver** [comment]
I'm not sure - where exactly user controls graph traversal pruning? :smiley:

**Comment by jdb** [comment]
Ok, "pruning" is cutting a whole area of the graph and we do not even look into it. The term is used in gardening as well, for real tree branches.

**Comment by jdb** [comment]
And in this solution, we have 2 parts: the traversal function that generates all the path and the functions that uses it to filter the path.

**Comment by jdb** [comment]
A traversal function that be communicated to stop recursing from the outside (from the caller) is what I find interesting

**Comment by jdb** [comment]
"Pruning a graph traversal can be controlled from the caller" if this is a bit easier to understand.

**Comment by jdb** [comment]
I hope I am not confusing you too much with yield and all that, maybe I will send you a solution without yield at all.

**Comment by SilverSolver** [comment]
No problem! I think I've understand the "yield" part here.
Thanks for explanation, it is more clear now what did you mean

File: jdb/assignment3/wordsearch.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    grid = Grid('abc', 'def', 'ghi', lexicon=True)
+    for letter, position in zip('abcdefghi',
+                                ((x, y) for y in range(3) for x in range(3))):
+      self.assertEqual(grid[position], letter)
+
+  def test_words(self):
+    words = Grid('aar', 'tcd').words()
+    for word in ['card', 'data', 'act', 'arc', 'cad', 'car', 'cat', 'rat',
+                 'rca', 'tad', 'tar', 'ac', 'ad']:
+      self.assertIn(word, words)
```
Comment (SilverSolver) [inline jdb/assignment3/wordsearch.py:134]: :slightly_frowning_face: Unittest give the following error:

**FAIL: test_words (wordsearch.TestGrid)
Traceback (most recent call last):
  File "/mnt/FileStorage/Projects_progs/ForVacancies/CodeU-2018-Group2/jdb/assignment3/wordsearch.py", line 138, in test_words
    self.assertIn(word, words)
AssertionError: 'data' not found in {'tad', 'cat', 'tc', 'tar', 'cr', 'act', 'rd', 'rca', 'card', 'ta', 'ar', 'car', 'arc', 'cd', 'cad', 'ad', 'ca', 'at', 'ac'}**

Then if I'm deleting the 'data', it shows the same error for 'rat'. The rest of words are found correctly.
What's wrong? :thinking: Both words 'data' and 'rat' contains diagonal step, but 'rca', for example, contains it too and not causing the error.

**Review by SilverSolver — CHANGES_REQUESTED** [review]
The last test cause an error

**Review by SilverSolver — APPROVED** [review]

**Pull request merged**